exporter/chronicleexporter/exporter.go

```diff
@@ -0,0 +1,131 @@
+)
+
+const scope = "https://www.googleapis.com/auth/malachite-ingestion"
+
+const baseEndpoint = "https://malachiteingestion-pa.googleapis.com"
+
+const apiTarget = "/v2/unstructuredlogentries:batchCreate"
+
+type chronicleExporter struct {
+	cfg        *Config
+	logger     *zap.Logger
+	httpClient *http.Client
+	marshaler  logMarshaler
+	endpoint   string
+}
+
+func newExporter(cfg *Config, params exporter.CreateSettings) (*chronicleExporter, error) {
+	var creds *google.Credentials
+	var err error
+
+	switch {
+	case cfg.Creds != "":
+		creds, err = google.CredentialsFromJSON(context.Background(), []byte(cfg.CredsFilePath), scope)
+		if err != nil {
+			return nil, fmt.Errorf("obtain credentials from JSON: %w", err)
+		}
+	case cfg.CredsFilePath != "":
+		credsData, err := os.ReadFile(cfg.CredsFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("read credentials file: %w", err)
+		}
+
+		creds, err = google.CredentialsFromJSON(context.Background(), credsData, scope)
+		if err != nil {
+			return nil, fmt.Errorf("obtain credentials from JSON: %w", err)
+		}
+	default:
+		return nil, fmt.Errorf("no credentials provided")
+	}
+
+	// Use the credentials to create an HTTP client
+	httpClient := oauth2.NewClient(context.Background(), creds.TokenSource)
+
+	return &chronicleExporter{
+		endpoint:   buildEndpoint(cfg),
+		cfg:        cfg,
+		logger:     params.Logger,
+		httpClient: httpClient,
+		marshaler:  newMarshaler(*cfg),
+	}, nil
+}
+
+// buildEndpoint builds the endpoint to send logs to based on the region. there is a default endpoint `https://malachiteingestion-pa.googleapis.com`
+// but there are also regional endpoints that can be used instead. the regional endpoints are listed here: https://cloud.google.com/chronicle/docs/reference/search-api#regional_endpoints
+func buildEndpoint(cfg *Config) string {
+	if cfg.Region != "" && regions[cfg.Region] != "" {
+		return fmt.Sprintf("%s%s", regions[cfg.Region], apiTarget)
+	}
+	return fmt.Sprintf("%s%s", baseEndpoint, apiTarget)
+}
+
+func (ce *chronicleExporter) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (ce *chronicleExporter) logsDataPusher(ctx context.Context, ld plog.Logs) error {
+	udmData, err := ce.marshaler.MarshalRawLogs(ld)
+	if err != nil {
+		return fmt.Errorf("marshal logs: %w", err)
+	}
+
+	return ce.uploadToChronicle(ctx, udmData)
+}
+
+func (ce *chronicleExporter) uploadToChronicle(ctx context.Context, data []byte) error {
+	request, err := http.NewRequestWithContext(ctx, "POST", ce.endpoint, bytes.NewBuffer(data))
+	if err != nil {
+		return fmt.Errorf("create request: %w", err)
+	}
+
+	request.Header.Set("Content-Type", "application/json")
+
+	resp, err := ce.httpClient.Do(request)
+	if err != nil {
+		return fmt.Errorf("send request to Chronicle: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		respBody := []byte{}
+		_, err := resp.Body.Read(respBody)
```

Comment:
Read() may not read the whole response, you'll need to read until EOF is encountered
```suggestion
		respBody, err := io.ReadAll(resp.Body)
```

Reply:
Updated.